### PR TITLE
Offer letter functionality

### PR DESCRIPTION
**Pull request by abhishek-pokhriyal**

### CODE REVIEW

File: resources/assets/js/app.js [hunk 1]
```diff
@@ -22,6 +22,21 @@ Vue.component('project-stage-component', require('./components/ProjectStageCompo
 Vue.component('project-stage-billing-component', require('./components/ProjectStageBillingComponent.vue'));
 Vue.component('applicant-round-action-component', require('./components/HR/ApplicantRoundActionComponent.vue'));
 
+$(document).ready(() => {
```
Comment (mohitgusain) [inline resources/assets/js/app.js]: why this change?

File: resources/views/finance/invoice/edit.blade.php [hunk 2]
```diff
@@ -14,7 +14,6 @@
         <form action="{{route('invoices.update', $invoice)}}" method="POST" enctype="multipart/form-data" class="form-invoice">
             @csrf
             @method('PATCH')
-
```
Comment (mohitgusain) [inline resources/views/finance/invoice/edit.blade.php]: remove extra line.

File: resources/assets/js/app.js (hunk 1 above)
Comment (abhishek-pokhriyal) [reply resources/assets/js/app.js]: only code position is changed. earlier this block was on line `168`.

File: resources/views/finance/invoice/edit.blade.php (hunk 2 above)
Comment (abhishek-pokhriyal) [reply resources/views/finance/invoice/edit.blade.php]: you mean add extra line. right?

File: app/Mail/HR/SendOfferLetter.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public $application;
+    public $body;
+    public $subject;
+
+    /**
+     * Create a new message instance.
+     *
+     * @return void
+     */
+    public function __construct(Application $application, $subject, $body)
```
Comment (rathorevaibhav) [inline app/Mail/HR/SendOfferLetter.php]: type-hint all function params.

File: app/Models/HR/Application.php [hunk 4]
```diff
@@ -382,8 +419,20 @@ public function isSentForApproval()
...
+    {
+        return $this->status == config('constants.hr.status.approved.label');
+    }
+
     public function isRejected()
     {
         return $this->status == config('constants.hr.status.rejected.label');
     }
+
+    public function saveOfferLetter(string $filePath)
```
Comment (rathorevaibhav) [inline app/Models/HR/Application.php]: you can avoid using this since the functionality within is also similar. We can avoid wrapping straightforward things.

File: app/Models/HR/ApplicationRound.php [hunk 5]
```diff
@@ -73,24 +77,67 @@ public function _update($attr)
             case 'send-for-approval':
                 $fillable['round_status'] = 'confirmed';
                 $application->sendForApproval($attr['send_for_approval_person']);
+
+                ApplicationMeta::create([
```
Comment (rathorevaibhav) [inline app/Models/HR/ApplicationRound.php]: It's not good to call one model from another model. Ideally a model's job should be just to interact with the specific resource it's supposed to take care of.

All this application logic like creating meta, sending emails, notifications etc should reside in a controller.

File: app/Models/HR/ApplicationRound.php [hunk 6]
```diff
@@ -73,24 +77,67 @@ public function _update($attr)
...
+                ]);
+
                 // The below env call needs to be changed to config after the default
                 // credentials bug in the Google API services is resolved.
                 $email = $attr['onboard_email'] . '@' . env('GOOGLE_CLIENT_HD');
                 $applicant->onboard($email, $attr['onboard_password'], [
                     'designation' => $attr['designation'],
                 ]);
+
+                User::create([
```
Comment (rathorevaibhav) [inline app/Models/HR/ApplicationRound.php]: This is supposed to happen after a certain event, i.e. when an applicant is onboarded. Ideally we should use Event/Listeners in Laravel. See if model observers can help here.

File: resources/views/hr/application/edit.blade.php [hunk 7]
```diff
@@ -183,10 +182,15 @@
                                 if ($loop->last) {
                                     if ($applicationRound->application->status == config('constants.hr.status.sent-for-approval.label')) {
```
Comment (rathorevaibhav) [inline resources/views/hr/application/edit.blade.php]: can merge first two if statements using `in_array`

File: resources/views/hr/application/index.blade.php [hunk 8]
```diff
@@ -79,6 +70,16 @@
                 @endif
                 </a>
             </li>
+            <li class="nav-item">
+                <a class="nav-item nav-link d-flex align-items-center {{ $status === config('constants.hr.status.sent-for-approval.label') ? 'active bg-info text-white' : 'text-info' }}" href= /{{ Request::path() .'?status='. config('constants.hr.status.sent-for-approval.label')}}{{request()->has('search')? "&search=".request('search'):"" }}
+               ><i class="fa fa-clock-o"></i>&nbsp;{{ config('constants.hr.status.sent-for-approval.title') }}
+                @if(request()->has('search'))
+                    <span class="ml-1 d-inline-block bg-info text-white px-2 py-0 {{ $status === config('constants.hr.status.sent-for-approval.label') ? 'active bg-white text-info' : 'text-white' }}" style="border-radius: 20px;font-size: 12px;font-weight: 700;">
```
Comment (rathorevaibhav) [inline resources/views/hr/application/index.blade.php]: `rounded`, `font-weight-bold`

Comment (rathorevaibhav) [reply resources/views/hr/application/index.blade.php]: remove inline css wherever you find as well.

File: resources/views/hr/application/offer-letter.blade.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE html>
```
Comment (rathorevaibhav) [inline resources/views/hr/application/offer-letter.blade.php]: isn't this supposed to be dynamic?

File: resources/views/hr/application/timeline.blade.php [hunk 10]
```diff
@@ -49,6 +46,28 @@
                             @include('hr.communication-mail-modal', ['data' => $event->communicationMail])
                         @endif
                         @break
+
+                    @case(config('constants.hr.application-meta.keys.sent-for-approval'))
+                        @php
+                            $event = $item['event'];
+                        @endphp
+                        <b><u>{{ date(config('constants.display_date_format'), strtotime($item['date'])) }}</u></b><br>
```
Comment (rathorevaibhav) [inline resources/views/hr/application/timeline.blade.php]: seems like you need this in all the three cases. Can this go out of the `switch` statement?

File: resources/views/settings/hr/approved.blade.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
+<div class="card mt-4">
+	<form action="{{ route('setting.hr.update') }}" method="POST">
+
```
Comment (rathorevaibhav) [inline resources/views/settings/hr/approved.blade.php]: remove extra lines

File: resources/views/settings/hr/approved.blade.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
+<div class="card mt-4">
```
Comment (rathorevaibhav) [inline resources/views/settings/hr/approved.blade.php]: rename this to approve

File: resources/views/settings/hr/index.blade.php [hunk 13]
```diff
@@ -10,6 +10,7 @@
     @include('settings.hr.applicant-auto-responder')
     @include('settings.hr.applicant-interview-reminder')
     @include('settings.hr.no-show')
+    @include('settings.hr.approved')
```
Comment (rathorevaibhav) [inline resources/views/settings/hr/index.blade.php]: hr.approve

File: routes/web.php [hunk 14]
```diff
@@ -62,9 +62,15 @@
 
         Route::prefix('applications')->namespace('Applications')->group(function () {
             Route::resource('/evaluation', 'EvaluationController')->only(['show', 'update']);
+
+            Route::get('job/{application}/offer-letter', 'JobApplicationController@viewOfferLetter')->name('applications.job.offer-letter');
```
Comment (rathorevaibhav) [inline routes/web.php]: did you test for all these resources? `job`, `internship` and `volunteer`? If not, please do the first round of self-testing

File: app/Models/HR/ApplicationRound.php (hunk 5 above)
Comment (abhishek-pokhriyal) [reply app/Models/HR/ApplicationRound.php]: We have opened to a new issue to track this #692

File: app/Models/HR/ApplicationRound.php (hunk 6 above)
Comment (abhishek-pokhriyal) [reply app/Models/HR/ApplicationRound.php]: #692
